In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn import model_selection
from sklearn.tree import DecisionTreeClassifier
from six import StringIO
from IPython.display import Image  
import graphviz
import pydotplus
from sklearn import metrics


In [2]:
dataN = pd.read_csv("dataN.csv")
dataN.head()

Unnamed: 0  mora  atraso  vivienda  edad  dias_lab  exp_sf  nivel_ahorro  \
0           0     0     235         1    30      3748    93.0             5   
1           1     0      18         1    32      4598     9.0            12   
2           2     0       0         1    26      5148     8.0             2   
3           3     0       0         1    36      5179    20.0            12   
4           4     0       0         1    46      3960     3.0             1   

   ingreso   linea_sf  deuda_sf  score         zona  clasif_sbs  nivel_educ  
0   3500.0  121543.40      0.00    214         Lima           4           3  
1    900.0    1824.67   1933.75    175  La Libertad           1           2  
2   2400.0    2797.38    188.29    187         Lima           0           3  
3   2700.0  121543.40      0.00    187       Ancash           0           2  
4   3100.0    2000.00  11010.65    189         Lima           0           2

In [3]:
feature_colsN = ['exp_sf', 'ingreso', 'score', 'dias_lab','nivel_ahorro']
Xn = dataN[feature_colsN] # features
yn = np.array(dataN['mora'])
Xn.shape

(8399, 5)

In [4]:
validation_size = 0.20
seed = 7
X_train, X_test, Y_train, Y_test = model_selection.train_test_split(Xn, yn, test_size=validation_size, random_state=seed)


# Árbol de decisión

In [5]:
'''
    Configuramos el max_depth = 2 para pre podar el árbol y asegurarse de que no tenga una profundidad mayor que 3. 
    También tenga en cuenta que en mi código a continuación, hice random_state = 0 para evitar la aleatoriedad
'''
# create Decision Tree classifier object
clf = DecisionTreeClassifier(criterion="entropy", max_depth=3, random_state = 0)

# train Decision Tree Classifier
clf = clf.fit(X_train,Y_train)

# predict the response for test dataset
y_pred = clf.predict(X_test)

# compute model accuracy
print("Accuracy:",metrics.accuracy_score(Y_test, y_pred))

Accuracy: 0.7297619047619047


## Importancia de la característica

In [6]:
#El siguiente código muestra las características importantes para cada característica en un modelo de árbol de decisión.

importances = pd.DataFrame({'feature':X_train.columns,'importance':np.round(clf.feature_importances_,3)})
importances = importances.sort_values('importance',ascending=False)
importances

feature  importance
0        exp_sf       0.586
4  nivel_ahorro       0.209
1       ingreso       0.205
2         score       0.000
3      dias_lab       0.000

## Visualización 

In [7]:
from sklearn.tree import export_text
from sklearn.model_selection import train_test_split

r = export_text(clf, feature_names=feature_colsN)

print(r)

|--- exp_sf <= 44.50
|   |--- ingreso <= 2096.80
|   |   |--- ingreso <= 826.30
|   |   |   |--- class: 1
|   |   |--- ingreso >  826.30
|   |   |   |--- class: 1
|   |--- ingreso >  2096.80
|   |   |--- nivel_ahorro <= 10.50
|   |   |   |--- class: 1
|   |   |--- nivel_ahorro >  10.50
|   |   |   |--- class: 1
|--- exp_sf >  44.50
|   |--- nivel_ahorro <= 11.50
|   |   |--- exp_sf <= 56.50
|   |   |   |--- class: 1
|   |   |--- exp_sf >  56.50
|   |   |   |--- class: 1
|   |--- nivel_ahorro >  11.50
|   |   |--- exp_sf <= 110.50
|   |   |   |--- class: 0
|   |   |--- exp_sf >  110.50
|   |   |   |--- class: 0



# SVM

In [8]:
#Librerías
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn import metrics

In [ ]:
# create a svm classifier with linear kernel
clf = svm.SVC(kernel='linear')

# train the model
clf.fit(X_train, Y_train)

# predict the response for the test dataset
y_pred = clf.predict(X_test)

In [ ]:
### Evaluate the model

# Model Accuracy: how often is the classifier correct?
print("Accuracy:", metrics.accuracy_score(Y_test, y_pred))

# Model Precision: what percentage of positive tuples are labeled as such?
print("Precision:", metrics.precision_score(Y_test, y_pred))

# Model Recall: what percentage of positive tuples are labelled as such?
print("Recall:", metrics.recall_score(Y_test, y_pred))

# Comparación precisiones

## Regresión logistica
Exactitud: 0.7148809523809524

## Árbol de decisión

Accuracy: 0.7297619047619047

## SVM
Accuracy: 0.7157142857142857